In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
from shapely.geometry import Point
from shapely.geometry import shape
import geopandas as gpd
from osgeo import gdal
import numpy as np
import os
import requests

In [2]:
cdl_codes = pd.read_csv("../unsynced-data/cdl-codes.csv")

In [3]:
#Load Data
counties = gpd.read_file("../unsynced-data/county-crops-v1.geojson")
counties = counties.to_crs("EPSG:4326")

In [49]:
cdl_codes = cdl_codes[cdl_codes.Class_Names != ' ']
cdl_codes = cdl_codes.reset_index(drop=True)
cdl_codes


Codes crop_code               Class_Names   category  Erdas_Red  \
0        0    crop_0                Background       meta      0.000   
1        1    crop_1                      Corn       food      1.000   
2        2    crop_2                    Cotton  commodity      1.000   
3        3    crop_3                      Rice       food      0.000   
4        4    crop_4                   Sorghum       food      1.000   
..     ...       ...                       ...        ...        ...   
129    247  crop_247                   Turnips       food      1.000   
130    248  crop_248                 Eggplants       food      1.000   
131    249  crop_249                    Gourds       food      1.000   
132    250  crop_250               Cranberries       food      1.000   
133    254  crop_254  Dbl Crop Barley/Soybeans       food      0.149   

     Erdas_Green  Erdas_Blue  ESRI_Red  ESRI_Green  ESRI_Blue  
0         0.0000      0.0000         0           0          0  
1         0.8275      0.0000       255         212          0  
2         0.1490      0.1490       255          38         38  
3         0.6588      0.8980         0         169        230  
4         0.6196      0.0471       255         158         15  
..           ...         ...       ...         ...        ...  
129       0.4000      0.4000       255         102        102  
130       0.4000      0.4000       255         102        102  
131       0.4000      0.4000       255         102        102  
132       0.4000      0.4000       255         102        102  
133       0.4431      0.0000        38         115          0  

[134 rows x 10 columns]

In [19]:
def c_to_crop(code):
    return cdl_codes[cdl_codes["Codes"] == code]["Class_Names"].values[0]
def cc_to_crop(crop_code):
    return cdl_codes[cdl_codes["crop_code"] == crop_code]["Class_Names"].values[0]

In [ ]:
col_list = []
for c in counties.columns:
    print(c)
    col_list.append(c)
crop_col = []
for c in col_list:
    if c.startswith("crop"):
        crop_col.append(c)

In [29]:
statefp = pd.read_csv("../unsynced-data/statefpcodes.csv")
# join State column and Postal column to counties based on FIPS and STATEFP
counties = counties.merge(statefp, left_on="STATEFP", right_on="FIPS")
counties = counties.drop(columns=["Unnamed: 3", "Unnamed: 4", "Unnamed: 5", "FIPS"])

In [37]:
counties["total_crop"] = counties[crop_col].sum(axis=1)
counties["max_crop_all"] = counties[crop_col].max(axis=1)

In [ ]:
# create a python list of all crop_codes that match category food in cdl_codes
food_crop = []
for i in range(len(cdl_codes)):
    if cdl_codes["category"][i] == "food":
        food_crop.append(cdl_codes["crop_code"][i])
# food_crop = [x for x in food_crop if x not in ['crop_210', 'crop_215', 'crop_228', 'crop_230', 'crop_234', 'crop_235']]
# counties["total_crop_food"] = counties[food_crop].sum(axis=1)
# counties["max_crop_food"] = counties[food_crop].max(axis=1)

In [97]:
counties.to_file("../unsynced-data/county-crops-v2.geojson", driver="GeoJSON")

In [93]:
counties

STATEFP COUNTYFP  COUNTYNS  GEOID        NAME           NAMELSAD LSAD  \
0         31      039  00835841  31039      Cuming      Cuming County   06   
1         31      109  00835876  31109   Lancaster   Lancaster County   06   
2         31      129  00835886  31129    Nuckolls    Nuckolls County   06   
3         31      101  00835872  31101       Keith       Keith County   06   
4         31      137  00835890  31137      Phelps      Phelps County   06   
...      ...      ...       ...    ...         ...                ...  ...   
2902      44      009  01219782  44009  Washington  Washington County   06   
2903      44      007  01219781  44007  Providence  Providence County   06   
2904      44      001  01219777  44001     Bristol     Bristol County   06   
2905      44      005  01219779  44005     Newport     Newport County   06   
2906      44      003  01219778  44003        Kent        Kent County   06   

     CLASSFP  MTFCC CSAFP  ... SOYBEANS_unit SUGARBEETS_unit SUGARCANE_unit  \
0         H1  G4020  None  ...             0               0              0   
1         H1  G4020   339  ...             0               0              0   
2         H1  G4020  None  ...             0               0              0   
3         H1  G4020  None  ...             0               0              0   
4         H1  G4020  None  ...             0               0              0   
...      ...    ...   ...  ...           ...             ...            ...   
2902      H4  G4020   148  ...             0               0              0   
2903      H4  G4020   148  ...             0               0              0   
2904      H4  G4020   148  ...             0               0              0   
2905      H4  G4020   148  ...             0               0              0   
2906      H4  G4020   148  ...             0               0              0   

      SUNFLOWER_unit  SWITCHGRASS_unit TOBACCO_unit TRITICALE_unit  \
0                  0                 0            0              0   
1                  0                 0            0              0   
2                  0                 0            0              0   
3                  0                 0            0              0   
4                  0                 0            0              0   
...              ...               ...          ...            ...   
2902               0                 0            0              0   
2903               0                 0            0              0   
2904               0                 0            0              0   
2905               0                 0            0              0   
2906               0                 0            0              0   

      WHEAT_unit  WILD RICE_unit  SWEET CORN_unit  
0              0               0                0  
1              0               0                0  
2              0               0                0  
3              0               0                0  
4              0               0                0  
...          ...             ...              ...  
2902           0               0                0  
2903           0               0                0  
2904           0               0                0  
2905           0               0                0  
2906           0               0                0  

[2907 rows x 240 columns]

In [67]:
kcal_conversion = pd.read_csv("../unsynced-data/Stability_Crop_Diversity-2.0/Data/Inputs/kcal_conversion.csv")
kcal_conversion

Crop   Crop_Class           Crop_Utilization  \
0    CAULIFLOWER  ALL CLASSES  ALL UTILIZATION PRACTICES   
1      ASPARAGUS  ALL CLASSES  ALL UTILIZATION PRACTICES   
2        LETTUCE         HEAD               FRESH MARKET   
3        LETTUCE         HEAD  ALL UTILIZATION PRACTICES   
4        LETTUCE         LEAF               FRESH MARKET   
..           ...          ...                        ...   
109       COTTON       UPLAND  ALL UTILIZATION PRACTICES   
110          HAY  ALL CLASSES  ALL UTILIZATION PRACTICES   
111         HOPS  ALL CLASSES  ALL UTILIZATION PRACTICES   
112      SORGHUM  ALL CLASSES                     SILAGE   
113      TOBACCO  ALL CLASSES  ALL UTILIZATION PRACTICES   

              Crop_Production Production_Units  Conversion_to_kg  kcal_kg  
0    ALL PRODUCTION PRACTICES              CWT          45.35920     98.0  
1    ALL PRODUCTION PRACTICES              CWT          45.35920    106.0  
2    ALL PRODUCTION PRACTICES              CWT          45.35920    114.0  
3    ALL PRODUCTION PRACTICES              CWT          45.35920    114.0  
4    ALL PRODUCTION PRACTICES              CWT          45.35920    140.0  
..                        ...              ...               ...      ...  
109  ALL PRODUCTION PRACTICES     480 LB BALES         217.68480      NaN  
110  ALL PRODUCTION PRACTICES             TONS         907.18500      NaN  
111  ALL PRODUCTION PRACTICES               LB           0.45351      NaN  
112  ALL PRODUCTION PRACTICES             TONS         907.18500      NaN  
113  ALL PRODUCTION PRACTICES               LB           0.45351      NaN  

[114 rows x 7 columns]

Made this request and downloaded all crop production data by county:

https://quickstats.nass.usda.gov/api/api_GET/?key=A5DB238B-8B04-3675-8A54-F373FA342D6F&source_desc=CENSUS&sector_desc=CROPS&agg_level_desc=COUNTY&domain_desc=TOTAL&statisticcat_desc=PRODUCTION&year__GE=2017&format=CSV

In [80]:
# all crop data by county in 2017
usda = pd.read_csv("../unsynced-data/usda_allcrops_county_2017.csv")
usda

group_desc commodity_desc   class_desc       prodn_practice_desc  \
0      CROP TOTALS    CROP TOTALS  ALL CLASSES       PRODUCTION CONTRACT   
1      CROP TOTALS    CROP TOTALS  ALL CLASSES       PRODUCTION CONTRACT   
2      CROP TOTALS    CROP TOTALS  ALL CLASSES       PRODUCTION CONTRACT   
3      CROP TOTALS    CROP TOTALS  ALL CLASSES       PRODUCTION CONTRACT   
4      CROP TOTALS    CROP TOTALS  ALL CLASSES       PRODUCTION CONTRACT   
...            ...            ...          ...                       ...   
46558   VEGETABLES     SWEET CORN  ALL CLASSES  ALL PRODUCTION PRACTICES   
46559   VEGETABLES     SWEET CORN  ALL CLASSES  ALL PRODUCTION PRACTICES   
46560   VEGETABLES     SWEET CORN  ALL CLASSES  ALL PRODUCTION PRACTICES   
46561   VEGETABLES     SWEET CORN  ALL CLASSES  ALL PRODUCTION PRACTICES   
46562   VEGETABLES     SWEET CORN  ALL CLASSES  ALL PRODUCTION PRACTICES   

              util_practice_desc   unit_desc  \
0      ALL UTILIZATION PRACTICES  OPERATIONS   
1      ALL UTILIZATION PRACTICES  OPERATIONS   
2      ALL UTILIZATION PRACTICES  OPERATIONS   
3      ALL UTILIZATION PRACTICES  OPERATIONS   
4      ALL UTILIZATION PRACTICES  OPERATIONS   
...                          ...         ...   
46558                       SEED          LB   
46559                       SEED          LB   
46560                       SEED          LB   
46561                       SEED          LB   
46562                       SEED          LB   

                                              short_desc  state_ansi  \
0      CROP TOTALS, PRODUCTION CONTRACT - OPERATIONS ...           1   
1      CROP TOTALS, PRODUCTION CONTRACT - OPERATIONS ...           4   
2      CROP TOTALS, PRODUCTION CONTRACT - OPERATIONS ...           5   
3      CROP TOTALS, PRODUCTION CONTRACT - OPERATIONS ...           6   
4      CROP TOTALS, PRODUCTION CONTRACT - OPERATIONS ...           6   
...                                                  ...         ...   
46558      SWEET CORN, SEED - PRODUCTION, MEASURED IN LB          53   
46559      SWEET CORN, SEED - PRODUCTION, MEASURED IN LB          53   
46560      SWEET CORN, SEED - PRODUCTION, MEASURED IN LB          53   
46561      SWEET CORN, SEED - PRODUCTION, MEASURED IN LB          55   
46562      SWEET CORN, SEED - PRODUCTION, MEASURED IN LB          55   

       state_fips_code state_alpha  state_name  asd_code             asd_desc  \
0                    1          AL     ALABAMA        60            WIREGRASS   
1                    4          AZ     ARIZONA        80             SOUTHERN   
2                    5          AR    ARKANSAS        60         EAST CENTRAL   
3                    6          CA  CALIFORNIA        40        CENTRAL COAST   
4                    6          CA  CALIFORNIA        80  SOUTHERN CALIFORNIA   
...                ...         ...         ...       ...                  ...   
46558               53          WA  WASHINGTON        50         EAST CENTRAL   
46559               53          WA  WASHINGTON        50         EAST CENTRAL   
46560               53          WA  WASHINGTON        50         EAST CENTRAL   
46561               55          WI   WISCONSIN        50              CENTRAL   
46562               55          WI   WISCONSIN        80        SOUTH CENTRAL   

       county_ansi  county_code      county_name  \
0             69.0           69          HOUSTON   
1             27.0           27             YUMA   
2             85.0           85           LONOKE   
3             79.0           79  SAN LUIS OBISPO   
4             65.0           65        RIVERSIDE   
...            ...          ...              ...   
46558          1.0            1            ADAMS   
46559         21.0           21         FRANKLIN   
46560         25.0           25            GRANT   
46561          1.0            1            ADAMS   
46562         25.0           25             DANE   

                                    location_desc    Va

In [88]:
# how many unique values in each column of usda
for c in usda.columns:
    print(c, len(usda[c].unique()))
    

group_desc 3
commodity_desc 48
class_desc 38
prodn_practice_desc 2
util_practice_desc 8
unit_desc 8
short_desc 87
state_ansi 50
state_fips_code 50
state_alpha 50
state_name 50
asd_code 23
asd_desc 86
county_ansi 267
county_code 269
county_name 1774
location_desc 3056
Value 24967
CV (%) 569


In [83]:
counties.columns

Index(['STATEFP', 'COUNTYFP', 'COUNTYNS', 'GEOID', 'NAME', 'NAMELSAD', 'LSAD',
       'CLASSFP', 'MTFCC', 'CSAFP',
       ...
       'crop_249', 'crop_250', 'crop_254', 'geometry', 'State', 'Postal',
       'total_crop', 'max_crop_all', 'total_crop_food', 'max_crop_food'],
      dtype='object', length=144)

In [84]:
# create a new column in counties for each unique value in the commodity_desc column
for i in range(len(usda["commodity_desc"].unique())):
    counties[usda["commodity_desc"].unique()[i]] = 0


In [86]:
counties.columns

Index(['STATEFP', 'COUNTYFP', 'COUNTYNS', 'GEOID', 'NAME', 'NAMELSAD', 'LSAD',
       'CLASSFP', 'MTFCC', 'CSAFP',
       ...
       'SOYBEANS', 'SUGARBEETS', 'SUGARCANE', 'SUNFLOWER', 'SWITCHGRASS',
       'TOBACCO', 'TRITICALE', 'WHEAT', 'WILD RICE', 'SWEET CORN'],
      dtype='object', length=192)

dummy check. columns went from 144 to 192. This lines up with the 48 unique values 

In [90]:
for i in range(len(usda["commodity_desc"].unique())):
    counties[usda["commodity_desc"].unique()[i]+"_unit"] = 0

In [ ]:
# for each of the rows in usda add the value of the value column to the column in counties that matches the commodity_desc in the row where usda county_code and counties COUNTYFP match
for i in range(len(usda)):
    if usda["county_code"][i] == counties["COUNTYFP"][i]:
        counties[usda["commodity_desc"][i]][i] = usda["value"][i]        

# Going with State Level Approach

1. Create DF with county crop pixels "unsynced-data/county-crops-v1.geojson"
2. Create a conversion dataframe with total pixels of each crop across the US with Crop Name
3. Use state level data to convert pixels to kCals "unsynced-data/Stability_Crop_Diversity-2.0/Data/Outputs/Intermediate_Data/Clean_Data.csv"
4. for every column in county level crops create a column with the kCals of that crop

In [4]:
counties = gpd.read_file("../unsynced-data/county-crops-v1.geojson")
counties = counties.to_crs("EPSG:4326")
for c in counties.columns:
    if c.startswith("crop"):
        counties[c] = counties[c].replace(0, np.nan)

production = pd.read_csv("../unsynced-data/Stability_Crop_Diversity-2.0/Data/Outputs/Intermediate_Data/Clean_Data.csv")
production = production.drop(columns=["Unnamed: 0"])

statefp = pd.read_csv("../unsynced-data/statefpcodes.csv")
counties = counties.merge(statefp, left_on="STATEFP", right_on="FIPS")
counties = counties.drop(columns=["Unnamed: 3", "Unnamed: 4", "Unnamed: 5", "FIPS"])

final_crops = pd.read_csv("../unsynced-data/final_crops.csv")

In [ ]:
# creates a dataframe with each unique crop name and the data associated with a row that has that crop name
# this was used to make the final_crops.csv file that will be all the crops we use in the final data
unique = production['Crop_Name'].unique().tolist()
unique_df = pd.DataFrame()
for u in unique:
    unique_df = unique_df.append(production[production['Crop_Name'] == u].iloc[0])
# write unique_df to csv
# unique_df.to_csv("../unsynced-data/Production_data_clean.csv")

Developing a function that will add kcal's to counties df
1. step 1 rollup counties data when needed as shown in final_crops dataframe
2. delete all columns that aren't present in the final_crops data
3. choose the closest year to 2017 or most recent in the production data
4. roll up values in production data when needed as shown in final_crops dataframe

In [12]:
for c in counties.columns:
    if c.startswith("crop"):
        counties[c] = counties[c].replace(0, np.nan)

In [13]:
counties

STATEFP COUNTYFP  COUNTYNS  GEOID        NAME           NAMELSAD LSAD  \
0         31      039  00835841  31039      Cuming      Cuming County   06   
1         31      109  00835876  31109   Lancaster   Lancaster County   06   
2         31      129  00835886  31129    Nuckolls    Nuckolls County   06   
3         31      101  00835872  31101       Keith       Keith County   06   
4         31      137  00835890  31137      Phelps      Phelps County   06   
...      ...      ...       ...    ...         ...                ...  ...   
2902      44      009  01219782  44009  Washington  Washington County   06   
2903      44      007  01219781  44007  Providence  Providence County   06   
2904      44      001  01219777  44001     Bristol     Bristol County   06   
2905      44      005  01219779  44005     Newport     Newport County   06   
2906      44      003  01219778  44003        Kent        Kent County   06   

     CLASSFP  MTFCC CSAFP  ... crop_245 crop_246 crop_247  crop_248  crop_249  \
0         H1  G4020  None  ...      NaN      NaN      NaN       NaN       NaN   
1         H1  G4020   339  ...      NaN      NaN      NaN       NaN       NaN   
2         H1  G4020  None  ...      NaN      NaN      NaN       NaN       NaN   
3         H1  G4020  None  ...      NaN      NaN      NaN       NaN       NaN   
4         H1  G4020  None  ...      NaN      NaN      NaN       NaN       NaN   
...      ...    ...   ...  ...      ...      ...      ...       ...       ...   
2902      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2903      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2904      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2905      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2906      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   

     crop_250 crop_254                                           geometry  \
0         NaN      NaN  MULTIPOLYGON (((-96.55515 41.91587, -96.55515 ...   
1         NaN      NaN  MULTIPOLYGON (((-96.68140 41.04566, -96.68139 ...   
2         NaN      NaN  MULTIPOLYGON (((-98.04802 40.35066, -98.04674 ...   
3         NaN      NaN  MULTIPOLYGON (((-101.73016 41.39521, -101.7263...   
4         NaN      NaN  MULTIPOLYGON (((-99.17949 40.52501, -99.17948 ...   
...       ...      ...                                                ...   
2902      NaN      NaN  MULTIPOLYGON (((-71.57520 41.32094, -71.57525 ...   
2903      NaN      NaN  MULTIPOLYGON (((-71.54735 41.73120, -71.54743 ...   
2904      NaN      NaN  MULTIPOLYGON (((-71.21043 41.68801, -71.21086 ...   
2905     32.0      NaN  MULTIPOLYGON (((-71.11640 41.48457, -71.11543 ...   
2906    362.0      NaN  MULTIPOLYGON (((-71.54481 41.60199, -71.54501 ...   

             State  Postal  
0         Nebraska      NE  
1         Nebraska      NE  
2         Nebraska      NE  
3         Nebraska      NE  
4         Nebraska      NE  
...            ...     ...  
2902  Rhode Island      RI  
2903  Rhode Island      RI  
2904  Rhode Island      RI  
2905  Rhode Island      RI  
2906  Rhode Island      RI  

[2907 rows x 137 columns]

In [22]:
production

State_Abbr  Year               Crop_Name  Price_Received_USD_kg  \
0             CA  1998    TOMATOES-ALL CLASSES               0.106907   
1             CA  1999     CARROTS-ALL CLASSES               0.343009   
2             CA  1999    TOMATOES-ALL CLASSES               0.097394   
3             CA  2000     CARROTS-ALL CLASSES               0.266750   
4             CA  2000    TOMATOES-ALL CLASSES               0.099191   
...          ...   ...                     ...                    ...   
22532         WY  2020      BARLEY-ALL CLASSES               0.226433   
22533         WY  2020  CORN-ALL CLASSES-GRAIN               0.155504   
22534         WY  2020         HAY-ALL CLASSES               0.188495   
22535         WY  2020  SUGARBEETS-ALL CLASSES               0.052690   
22536         WY  2020       WHEAT-ALL CLASSES               0.173063   

        Crop_Area_ha  Production_kg  FIPS  kcal_kg  Production_kcal  \
0      130716.309187   8.535522e+09     6    187.0     1.596143e+12   
1       30109.267503   1.029173e+09     6    365.0     3.756482e+11   
2      152165.115338   1.165760e+10     6    187.0     2.179971e+12   
3       29906.920275   1.025345e+09     6    365.0     3.742509e+11   
4      127478.753541   9.910542e+09     6    187.0     1.853271e+12   
...              ...            ...   ...      ...              ...   
22532   27114.528531   1.400401e+08    56   3530.0     4.943415e+11   
22533   21853.500607   1.673431e+08    56   3650.0     6.108023e+11   
22534  437070.012141   2.440328e+09    56      NaN              NaN   
22535   12383.650344   8.219096e+08    56    700.0     5.753367e+11   
22536   36422.501012   6.368427e+07    56   3338.0     2.125781e+11   

       Production_USD  
0        1.317294e+09  
1        5.417210e+08  
2        1.742295e+09  
3        4.150759e+08  
4        1.491845e+09  
...               ...  
22532    3.032409e+07  
22533    2.488545e+07  
22534    4.398891e+08  
22535    4.141436e+07  
22536    1.053978e+07  

[22537 rows x 10 columns]

In [47]:
final_crops

crop_code        cdl_name                   production_name  \
0    crop_75         Almonds               ALMONDS-ALL CLASSES   
1    crop_68          Apples                APPLES-ALL CLASSES   
2   crop_223        Apricots              APRICOTS-ALL CLASSES   
3   crop_207       Asparagus             ASPARAGUS-ALL CLASSES   
4   crop_215        Avocados              AVOCADOS-ALL CLASSES   
5    crop_21          Barley                BARLEY-ALL CLASSES   
6    crop_42       Dry Beans  BEANS-DRY EDIBLE, INCL CHICKPEAS   
7   crop_242     Blueberries                  BLUEBERRIES-TAME   
8   crop_214        Broccoli              BROCCOLI-ALL CLASSES   
9   crop_243         Cabbage               CABBAGE-ALL CLASSES   
10   crop_31          Canola                CANOLA-ALL CLASSES   
11  crop_206         Carrots               CARROTS-ALL CLASSES   
12  crop_244     Cauliflower           CAULIFLOWER-ALL CLASSES   
13  crop_245          Celery                CELERY-ALL CLASSES   
14   crop_66        Cherries                    CHERRIES-SWEET   
15    crop_1            Corn            CORN-ALL CLASSES-GRAIN   
16  crop_250     Cranberries           CRANBERRIES-ALL CLASSES   
17   crop_50       Cucumbers             CUCUMBERS-ALL CLASSES   
18  crop_248       Eggplants              EGGPLANT-ALL CLASSES   
19   crop_32        Flaxseed              FLAXSEED-ALL CLASSES   
20  crop_208          Garlic                GARLIC-ALL CLASSES   
21   crop_69          Grapes                GRAPES-ALL CLASSES   
22   crop_52         Lentils               LENTILS-ALL CLASSES   
23  crop_227         Lettuce                      LETTUCE-HEAD   
24  crop_209     Cantaloupes                  MELONS-CANTALOUP   
25   crop_48     Watermelons                 MELONS-WATERMELON   
26   crop_29          Millet                      MILLET-PROSO   
27   crop_35         Mustard               MUSTARD-ALL CLASSES   
28  crop_218      Nectarines            NECTARINES-ALL CLASSES   
29   crop_28            Oats                  OATS-ALL CLASSES   
30  crop_211          Olives                OLIVES-ALL CLASSES   
31   crop_49          Onions                        ONIONS-DRY   
32  crop_212         Oranges               ORANGES-ALL CLASSES   
33   crop_67         Peaches               PEACHES-ALL CLASSES   
34   crop_10         Peanuts               PEANUTS-ALL CLASSES   
35   crop_77           Pears                 PEARS-ALL CLASSES   
36   crop_53            Peas                        PEAS-GREEN   
37   crop_74          Pecans                PECANS-ALL CLASSES   
38  crop_216         Peppers                     PEPPERS-CHILE   
39  crop_204      Pistachios            PISTACHIOS-ALL CLASSES   
40  crop_220           Plums                 PLUMS-ALL CLASSES   
41   crop_43        Potatoes              POTATOES-ALL CLASSES   
42  crop_210          Prunes                PRUNES-ALL CLASSES   
43  crop_229        Pumpkins              PUMPKINS-ALL CLASSES   
44  crop_246        Radishes              RADISHES-ALL CLASSES   
45    crop_3            Rice                  RICE-ALL CLASSES   
46   crop_27             Rye                   RYE-ALL CLASSES   
47   crop_33       Safflower             SAFFLOWER-ALL CLASSES   
48    crop_4         Sorghum         SORGHUM-ALL CLASSES-GRAIN   
49    crop_5        Soybeans              SOYBEANS-ALL CLASSES   
50  crop_222          Squash                SQUASH-ALL CLASSES   
51  crop_221    Strawberries          STRAWBERRIES-ALL CLASSES   
52   crop_41      Sugarbeets            SUGARBEETS-ALL CLASSES   
53    crop_6       Sunflower             SUNFLOWER-ALL CLASSES   
54   crop_12      Sweet Corn            SWEET CORN-ALL CLASSES   
55   crop_46  Sweet Potatoes        SWEET POTATOES-ALL CLASSES   
56   crop_54        Tomatoes              TOMATOES-ALL CLASSES   
57   crop_76         Walnuts                   WALNUTS-ENGLISH   
58   crop_24    Winter Wheat                 WHEAT-ALL CLASSES   
59   crop_72          Citrus            GRA

In [6]:
# rolling up counties data as dictated by final crops
counties["crop_42"] = counties["crop_42"] + counties["crop_51"]
counties["crop_22"] = counties["crop_22"] + counties["crop_23"] + counties["crop_24"]
counties = counties.drop(columns=["crop_51", "crop_23", "crop_24"])

In [ ]:
# rolling up production data as dictated by final crops
# this basically takes each crop from each state and adds the row that is closest to 2017 to the rel_prod dataframe
rel_prod = pd.DataFrame()
for s in production["State_Abbr"].unique():
    state = production[production["State_Abbr"] == s]
    for c in state["Crop_Name"].unique():
        crop = state[state["Crop_Name"] == c]
        rel_prod = rel_prod.append(crop.iloc[(crop['Year']-2017).abs().argsort()[:1]])

In [ ]:
# joining rollup rows on production data
roll1 = ["LETTUCE-HEAD", "LETTUCE-LEAF", "LETTUCE-ROMAINE"]
roll1_df = pd.DataFrame()
roll1_df = roll1_df.append(rel_prod.iloc[0])
roll1_df = roll1_df.drop(roll1_df.index[0])
roll1_df
for s in rel_prod["State_Abbr"].unique():
    state = rel_prod[rel_prod["State_Abbr"] == s]
    # make a data frame of values that match roll1 and state
    temp = state[state["Crop_Name"].isin(roll1)]
    # add a row to roll1_df that includes state abbreviation, most recent year, "LETTUCE-HEAD", average Price_Received_USD_kg, sum of values in in Crop_Area_ha, sum of values in Production_kg, FIPS in FIPS column, 138 kcal_kg in kcal_kg column, sum of values in Production_kcal, sum of values in Production_USD columns
    # if temp length is longer than 0
    if len(temp) > 0:
        roll1_df = roll1_df.append({
            "State_Abbr": s, "Year": temp["Year"].max(), 
            "Crop_Name": "LETTUCE-HEAD", 
            "Price_Received_USD_kg": temp["Price_Received_USD_kg"].mean(), 
            "Crop_Area_ha": temp["Crop_Area_ha"].sum(), "Production_kg": temp["Production_kg"].sum(), 
            "FIPS": temp["FIPS"].iloc[0], 
            "kcal_kg": temp["kcal_kg"].mean(),
            "Production_kcal": temp["Production_kcal"].sum(), 
            "Production_USD": temp["Production_USD"].sum()
            }, ignore_index=True)
roll1_df = roll1_df.drop_duplicates(subset="State_Abbr", keep="first")
for i in range(len(roll1)):
    rel_prod = rel_prod[rel_prod["Crop_Name"] != roll1[i]]
rel_prod = rel_prod.append(roll1_df)

roll2 = ["PEAS-GREEN", "PEAS-DRY EDIBLE", "PEAS-AUSTRIAN WINTER"]
roll2_df = pd.DataFrame()
roll2_df = roll2_df.append(rel_prod.iloc[0])
roll2_df = roll2_df.drop(roll2_df.index[0])
for s in rel_prod["State_Abbr"].unique():
    state = rel_prod[rel_prod["State_Abbr"] == s]
    temp = state[state["Crop_Name"].isin(roll2)]
    if len(temp) > 0:
        roll2_df = roll2_df.append({
            "State_Abbr": s, "Year": temp["Year"].max(), 
            "Crop_Name": "PEAS-GREEN", 
            "Price_Received_USD_kg": temp["Price_Received_USD_kg"].mean(), 
            "Crop_Area_ha": temp["Crop_Area_ha"].sum(), "Production_kg": temp["Production_kg"].sum(), 
            "FIPS": temp["FIPS"].iloc[0], 
            "kcal_kg": temp["kcal_kg"].mean(), 
            "Production_kcal": temp["Production_kcal"].sum(), 
            "Production_USD": temp["Production_USD"].sum()
            }, ignore_index=True)
roll2_df = roll2_df.drop_duplicates(subset="State_Abbr", keep="first")
for i in range(len(roll2)):
    rel_prod = rel_prod[rel_prod["Crop_Name"] != roll2[i]]
rel_prod = rel_prod.append(roll2_df)

roll3 = ["GRAPEFRUIT-ALL CLASSES", "LEMONS-ALL CLASSES", "TANGELOS-ALL CLASSES", "TANGERINES-ALL CLASSES"]
roll3_df = pd.DataFrame()
roll3_df = roll3_df.append(rel_prod.iloc[0])
roll3_df = roll3_df.drop(roll3_df.index[0])
for s in rel_prod["State_Abbr"].unique():
    state = rel_prod[rel_prod["State_Abbr"] == s]
    temp = state[state["Crop_Name"].isin(roll3)]
    if len(temp) > 0:
        roll3_df = roll3_df.append({
            "State_Abbr": s, "Year": temp["Year"].max(), 
            "Crop_Name": "GRAPEFRUIT-ALL CLASSES", 
            "Price_Received_USD_kg": temp["Price_Received_USD_kg"].mean(), 
            "Crop_Area_ha": temp["Crop_Area_ha"].sum(), "Production_kg": temp["Production_kg"].sum(), 
            "FIPS": temp["FIPS"].iloc[0], 
            "kcal_kg": temp["kcal_kg"].mean(), 
            "Production_kcal": temp["Production_kcal"].sum(), 
            "Production_USD": temp["Production_USD"].sum()
            }, ignore_index=True)
roll3_df = roll3_df.drop_duplicates(subset="State_Abbr", keep="first")
for i in range(len(roll3)):
    rel_prod = rel_prod[rel_prod["Crop_Name"] != roll3[i]]
rel_prod = rel_prod.append(roll3_df)
rel_prod = rel_prod.reset_index(drop=True)

In [11]:
rel_prod

State_Abbr  Year                         Crop_Name  Price_Received_USD_kg  \
0           CA  2017              TOMATOES-ALL CLASSES               0.106924   
1           CA  2017               CARROTS-ALL CLASSES               0.648160   
2           CA  2017                BARLEY-ALL CLASSES               0.221381   
3           CA  2017  BEANS-DRY EDIBLE, INCL CHICKPEAS               1.349230   
4           CA  2017            CORN-ALL CLASSES-GRAIN               0.165347   
..         ...   ...                               ...                    ...   
918         SD  2017                        PEAS-GREEN               0.246918   
919         CA  2017            GRAPEFRUIT-ALL CLASSES               0.569895   
920         FL  2017            GRAPEFRUIT-ALL CLASSES               0.469951   
921         TX  2017            GRAPEFRUIT-ALL CLASSES               0.357700   
922         AZ  2017            GRAPEFRUIT-ALL CLASSES               0.346838   

      Crop_Area_ha  Production_kg  FIPS      kcal_kg  Production_kcal  \
0    100242.816673   1.005575e+10     6   187.000000     1.880425e+12   
1     23674.625658   9.817999e+08     6   365.000000     3.583570e+11   
2     11736.139215   3.156998e+07     6  3530.000000     1.114420e+11   
3     20113.314448   4.740036e+07     6  3506.000000     1.661857e+11   
4     32375.556455   3.393600e+08     6  3650.000000     1.238664e+12   
..             ...            ...   ...          ...              ...   
918   14164.305949   2.381358e+07    46  2620.000000     6.239158e+10   
919   46701.740186   1.767196e+09     6   280.666667     5.798218e+11   
920   18494.536625   3.858034e+08    12   314.666667     8.182288e+10   
921    6636.989073   1.741795e+08    48   160.000000     2.786872e+10   
922    3601.780656   6.332151e+07     4   280.666667     1.887453e+10   

     Production_USD  
0      1.003435e+09  
1      5.938864e+08  
2      6.522491e+06  
3      5.968513e+07  
4      5.236658e+07  
..              ...  
918    5.487515e+06  
919    5.744850e+08  
920    1.456472e+08  
921    5.814527e+07  
922    2.837039e+06  

[923 rows x 10 columns]

Developing pixel to kcal conversion function. I'm going to develop a national level that will take total national pixels adn 

In [ ]:
# in counties drop any column whose column name starts with "crop" and is not in final_crop[crop_code] column
# this is to remove any columns that are not in the final crop list
rel_counties = counties
for c in counties.columns:
    if c.startswith("crop") and c not in final_crops["crop_code"].values:
        rel_counties = rel_counties.drop(c, axis=1)
rel_counties

In [28]:
counties["crop_75"].sum()

0.0

In [26]:
for c in counties.columns:
    if c.startswith("crop"):
        final_crops.loc[final_crops["crop_code"] == c, "total_pixels"] = counties[c].sum()

In [30]:
rel_counties

STATEFP COUNTYFP  COUNTYNS  GEOID        NAME           NAMELSAD LSAD  \
0         31      039  00835841  31039      Cuming      Cuming County   06   
1         31      109  00835876  31109   Lancaster   Lancaster County   06   
2         31      129  00835886  31129    Nuckolls    Nuckolls County   06   
3         31      101  00835872  31101       Keith       Keith County   06   
4         31      137  00835890  31137      Phelps      Phelps County   06   
...      ...      ...       ...    ...         ...                ...  ...   
2902      44      009  01219782  44009  Washington  Washington County   06   
2903      44      007  01219781  44007  Providence  Providence County   06   
2904      44      001  01219777  44001     Bristol     Bristol County   06   
2905      44      005  01219779  44005     Newport     Newport County   06   
2906      44      003  01219778  44003        Kent        Kent County   06   

     CLASSFP  MTFCC CSAFP  ... crop_242 crop_243 crop_244  crop_245  crop_246  \
0         H1  G4020  None  ...      NaN     11.0      NaN       NaN       NaN   
1         H1  G4020   339  ...      NaN      NaN      NaN       NaN       NaN   
2         H1  G4020  None  ...      NaN      NaN      NaN       NaN       NaN   
3         H1  G4020  None  ...      NaN      NaN      NaN       NaN       NaN   
4         H1  G4020  None  ...      NaN      NaN      NaN       NaN       NaN   
...      ...    ...   ...  ...      ...      ...      ...       ...       ...   
2902      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2903      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2904      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2905      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   
2906      H4  G4020   148  ...      NaN      NaN      NaN       NaN       NaN   

     crop_248 crop_250                                           geometry  \
0         NaN      NaN  MULTIPOLYGON (((-96.55515 41.91587, -96.55515 ...   
1         NaN      NaN  MULTIPOLYGON (((-96.68140 41.04566, -96.68139 ...   
2         NaN      NaN  MULTIPOLYGON (((-98.04802 40.35066, -98.04674 ...   
3         NaN      NaN  MULTIPOLYGON (((-101.73016 41.39521, -101.7263...   
4         NaN      NaN  MULTIPOLYGON (((-99.17949 40.52501, -99.17948 ...   
...       ...      ...                                                ...   
2902      NaN      NaN  MULTIPOLYGON (((-71.57520 41.32094, -71.57525 ...   
2903      NaN      NaN  MULTIPOLYGON (((-71.54735 41.73120, -71.54743 ...   
2904      NaN      NaN  MULTIPOLYGON (((-71.21043 41.68801, -71.21086 ...   
2905      NaN     32.0  MULTIPOLYGON (((-71.11640 41.48457, -71.11543 ...   
2906      NaN    362.0  MULTIPOLYGON (((-71.54481 41.60199, -71.54501 ...   

             State  Postal  
0         Nebraska      NE  
1         Nebraska      NE  
2         Nebraska      NE  
3         Nebraska      NE  
4         Nebraska      NE  
...            ...     ...  
2902  Rhode Island      RI  
2903  Rhode Island      RI  
2904  Rhode Island      RI  
2905  Rhode Island      RI  
2906  Rhode Island      RI  

[2907 rows x 77 columns]

In [35]:
rel_prod.head(5)

State_Abbr  Year                         Crop_Name  Price_Received_USD_kg  \
0         CA  2017              TOMATOES-ALL CLASSES               0.106924   
1         CA  2017               CARROTS-ALL CLASSES               0.648160   
2         CA  2017                BARLEY-ALL CLASSES               0.221381   
3         CA  2017  BEANS-DRY EDIBLE, INCL CHICKPEAS               1.349230   
4         CA  2017            CORN-ALL CLASSES-GRAIN               0.165347   

    Crop_Area_ha  Production_kg  FIPS  kcal_kg  Production_kcal  \
0  100242.816673   1.005575e+10     6    187.0     1.880425e+12   
1   23674.625658   9.817999e+08     6    365.0     3.583570e+11   
2   11736.139215   3.156998e+07     6   3530.0     1.114420e+11   
3   20113.314448   4.740036e+07     6   3506.0     1.661857e+11   
4   32375.556455   3.393600e+08     6   3650.0     1.238664e+12   

   Production_USD  
0    1.003435e+09  
1    5.938864e+08  
2    6.522491e+06  
3    5.968513e+07  
4    5.236658e+07

In [ ]:
# for every unique value in "Crop_Name" add the values from those, and add them to the final_crops dataframe
for c in rel_prod["Crop_Name"].unique():
    final_crops.loc[final_crops["production_name"] == c, "Crop_Area_ha"] = rel_prod[rel_prod["Crop_Name"] == c]["Crop_Area_ha"].sum()
    final_crops.loc[final_crops["production_name"] == c, "Production_kg"] = rel_prod[rel_prod["Crop_Name"] == c]["Production_kg"].sum()
    final_crops.loc[final_crops["production_name"] == c, "Production_kcal"] = rel_prod[rel_prod["Crop_Name"] == c]["Production_kcal"].sum()
    final_crops.loc[final_crops["production_name"] == c, "Production_USD"] = rel_prod[rel_prod["Crop_Name"] == c]["Production_USD"].sum()
    final_crops.loc[final_crops["production_name"] == c, "Price_Received_USD_kg"] = rel_prod[rel_prod["Crop_Name"] == c]["Price_Received_USD_kg"].mean()

In [39]:
final_crops["kcal_pixel_national"] = final_crops["Production_kcal"] / final_crops["total_pixels"]

In [42]:
# write final_crops to csv
final_crops.to_csv("final_crops_with_kcal.csv")